src/runtime/FinalizationRegistryObject.cpp

```diff
@@ -73,4 +90,96 @@ bool FinalizationRegistryObject::deleteCell(ExecutionState& state, Value unregis
     return removed;
 }
 
+class CleanupSomeJob : public Job {
+public:
+    CleanupSomeJob(Context* relatedContext, FinalizationRegistryObject* object, Optional<Object*> callback)
+        : Job(relatedContext)
+        , m_object(object)
+        , m_callback(callback)
+    {
+    }
+
+    SandBox::SandBoxResult run()
+    {
+        auto oldCallback = m_object->m_cleanupCallback;
+        m_object->m_cleanupCallback = m_callback;
+
+        clearStack<1024>();
+        GC_gcollect_and_unmap();
+        GC_gcollect_and_unmap();
+        GC_gcollect_and_unmap();
+        m_object->m_cleanupCallback = oldCallback;
+
+        SandBox sandbox(relatedContext());
+        ExecutionState state(relatedContext());
+        return sandbox.run([&]() -> Value {
+            return Value();
+        });
+    }
+
+private:
+    FinalizationRegistryObject* m_object;
+    Optional<Object*> m_callback;
+};
+
+void FinalizationRegistryObject::cleanupSome(ExecutionState& state, Optional<Object*> callback)
+{
+    state.context()->vmInstance()->jobQueue()->enqueueJob(new CleanupSomeJob(state.context(), this, callback));
+}
+
+void* FinalizationRegistryObject::FinalizationRegistryObjectItem::operator new(size_t size)
+{
+#ifdef NDEBUG
+    static bool typeInited = false;
+    static GC_descr descr;
+    if (!typeInited) {
+        GC_word obj_bitmap[GC_BITMAP_SIZE(FinalizationRegistryObject::FinalizationRegistryObjectItem)] = { 0 };
+        GC_set_bit(obj_bitmap, GC_WORD_OFFSET(FinalizationRegistryObject::FinalizationRegistryObjectItem, source));
```

Comment:
One more question.
`FinalizationRegistryObjectItem` holds `source`(FinalizationRegistryObject) member which means that `FinalizationRegistryObject` will be valid as long as WeakRefTarget is alive. 
According to the standard, is it impossible case that `FinalizationRegistryObject` is already gc-collected but WeakRefTarget is still alive?

Reply:
bdwgc cannot remove dangling objects at once.
FinalizationRegistryObjectItem can alive but FinalizationRegistryObject can be killed by bc
but, finalizer references source member variable.
that's why I want to track source

Reply:
Then the following situation is also possible?
`FinalizationRegistryObject` and `WeakRefTarget` are no longer valid at the same time.
Let's assume that gc removes `FinalizationRegistryObject` first and then `WeakRefTarget`.
In this case, registered finalizer callback of `WeakRefTarget` will not be called.
Is this correct?

Reply:
yes, we remove WeakRefTarget's  finalizer at finalizer in FinalizationRegistryObject 
you are right!